src/data/fin.lean

```diff
@@ -98,6 +98,43 @@ attribute [simp] val_zero
 @[simp] lemma coe_one  {n : ℕ} : ((1 : fin (n+2)) : ℕ) = 1 := rfl
 @[simp] lemma coe_two  {n : ℕ} : ((2 : fin (n+3)) : ℕ) = 2 := rfl
 
+/-- `a < b` as natural numbers if and only if `a < b` in `fin n`. -/
+@[norm_cast, simp] lemma coe_fin_lt {n : ℕ} {a b : fin n} : (a : ℕ) < (b : ℕ) ↔ a < b :=
+iff.rfl
+
+/-- `a ≤ b` as natural numbers if and only if `a ≤ b` in `fin n`. -/
+@[norm_cast, simp] lemma coe_fin_le {n : ℕ} {a b : fin n} : (a : ℕ) ≤ (b : ℕ) ↔ a ≤ b :=
+iff.rfl
+
+/-- Converting an in-range number to `fin (n + 1)` with `of_nat`
+produces a result whose value is the original number.  -/
+lemma of_nat_val_of_lt {n : ℕ} {a : ℕ} (h : a < n + 1) :
+  ((of_nat a) : fin (n + 1)).val = a :=
+nat.mod_eq_of_lt h
+
+/-- Converting the value of a `fin (n + 1)` to `fin (n + 1)` with `of_nat`
+results in the same value.  -/
+@[simp] lemma of_nat_val_eq_self {n : ℕ} (a : fin (n + 1)) : of_nat a.val = a :=
+begin
+  rw eq_iff_veq,
+  exact of_nat_val_of_lt a.is_lt
+end
+
+/-- `of_nat` of an in-range number, converted back to `ℕ`, is that
+number. -/
+lemma of_nat_coe_of_lt {n : ℕ} {a : ℕ} (h : a < n + 1) :
+  (((of_nat a) : fin (n + 1)) : ℕ) = a :=
+nat.mod_eq_of_lt h
+
+/-- Converting a `fin (n + 1)` to `ℕ` and back with `of_nat` results in
+the same value. -/
+@[simp] lemma of_nat_coe_eq_self {n : ℕ} (a : fin (n + 1)) : of_nat (a : ℕ) = a :=
+of_nat_val_eq_self a
+
+/-- `of_nat 0`, converted to `ℕ`, is 0. -/
+@[simp] lemma of_nat_coe_zero {n : ℕ} : (((of_nat 0) : fin (n + 1)) : ℕ) = 0 :=
```

Comment:
I think that `(of_nat 0 : fin (n + 1)) = 0` would be a better `simp` lemma. Combined with the existing `coe_zero` it gives your lemma.

BTW, should `coe_zero`, `coe_one`, and `coe_two` be `norm_cast`?

Reply:
It seems that's `of_nat_zero` (in core, not mathlib). Does that mean this lemma should be replaced by just `attribute [simp] fin.of_nat_zero`? No idea whether those other lemmas should be `norm_cast`.

Reply:
Yes, and while you're at it, could you please add `@[simp] lemma of_nat_one`?

Reply:
Now done.